### PR TITLE
Harmbattons on harm intent will now properly stun people

### LINKED ISSUE #16838: Activated Stun Batons on the Harm Intent no Longer Stun
#### Description of issue
Activated Harmbatons seem to no longer stun when applied to onesself or others.

#### Difference between expected and actual behavior
Activated Harmbatons used to stun for a period in addition to dealing brute damage when applied to clowns/cultists/unsuspecting security officers.

#### Steps to reproduce 
1. activate a stun baton
2. enter harm intent
3. smack yourself in the chest with the baton
3a. (optional: smack a friend in the chest with the baton)
4. ponder your sustained verticality

#### Length of time in which bug has been known to occur
I first noticed it maybe a week or so ago, but I don't think anyone else knows about it.

#### Server revision
Game: 2bf9d817088ea972f55540db3315ded858803015 
Patches: 9e34bcea4d28270b966365362177f6890fc3ca45 

#### Issue bingo
Please check whatever applies. More checkboxes checked increase your chances of not being yelled at by every contributor.

- [x] Issue could be reproduced at least once
- [x] Issue could be reproduced by different players
(does getting someone on the test serb to harmbaton me count?)
- [x] Issue could be reproduced in multiple rounds
- [x] Issue happened in a recent (less than 7 days ago) round
- [x] [Couldn't find an existing issue about this](https://github.com/d3athrow/vgstation13/issues)

### PR DESCRIPTION
**Pull request by ShiftyRail**
Fixes #16838
Apparently `hit` was null, or improperly defined, which caused a silent runtime, I believe.
Tested

### CODE REVIEW

File: code/game/objects/items/weapons/stunbaton.dm [hunk 1]
```diff
@@ -161,7 +161,8 @@
 
 	var/hit = 1
 	if(user.a_intent == I_HURT)
-		hit = ..()
+		. = ..()
+		hit = -1
```
Comment (MadmanMartian) [inline code/game/objects/items/weapons/stunbaton.dm]: What does this being -1 do exactly?

Below, -1 is what is used for when the baton is off. Wouldn't this cause potential problems?

Comment (ShiftyRail) [reply code/game/objects/items/weapons/stunbaton.dm]: I _think_ it is used to do a check for miss later on.
There is no check for miss on help intent.
~This code is barely readable, I did this quickly and it needs moar testing~

**Comment by Accipitermaris** [comment]
Your fathers stun baton. This is the 
                         weapon of a Security Officer. Not as clumsy 
                         or as random as a taser.

Thank you Santa <3

**Comment by DamianX** [comment]
Does it also play the hitsound now?

File: code/game/objects/items/weapons/stunbaton.dm [hunk 2]
```diff
@@ -159,38 +161,32 @@
 
 	var/mob/living/L = M
 
-	var/hit = 1
```
Comment (ShiftyRail) [inline code/game/objects/items/weapons/stunbaton.dm]: This var did literally nothing.

File: code/game/objects/items/weapons/stunbaton.dm [hunk 3]
```diff
@@ -159,38 +161,32 @@
...
-	if(user.a_intent == I_HURT)
-		hit = ..()
-		if(hit)
-			playsound(loc, "swing_hit", 50, 1, -1)
-			user.do_attack_animation(M, src)
+	if(user.a_intent == I_HURT) // Harm intent : possibility to miss (in exchange for doing actual damage)
+		var/target_zone = get_zone_with_miss_chance(user.zone_sel.selecting, L)
+		if (user != L && !target_zone)
+			visible_message("<span class='danger'>\The [user] misses \the [L] with \the [src]!</span>")
+			playsound(loc, "sound/weapons/punchmiss.ogg", 50, 1, -1)
```
Comment (ShiftyRail) [inline code/game/objects/items/weapons/stunbaton.dm]: Feedback is nice to the player.

**Comment by ShiftyRail** [comment]
I ended up refactoring this a little bit.
Here's a summary of the changes, with explanations why (Inline comments were obnoxious for review)

- `hit` was just a check for if we had harm intent. Using magic numbers. I removed it and placed the check for miss in the body of the "harm intent" section.
- When you miss, you'll now be informed that you missed with a nice sound effect.

@DamianX Yeah, it plays a hitsound on hit and also when you miss.

File: code/game/objects/items/weapons/stunbaton.dm [hunk 4]
```diff
@@ -1,3 +1,5 @@
+#define HITTING_BUT_CAN_MISS -1 // Danke, old coders
```
Comment (MadmanMartian) [inline code/game/objects/items/weapons/stunbaton.dm]: If the var did nothing by itself, then there's no need for the define, right?

Comment (ShiftyRail) [reply code/game/objects/items/weapons/stunbaton.dm]: You saw nothing :^)

**Comment by ShiftyRail** [comment]
Wait I think I forgot something : hitting someone restrained/lying on the ground should always work.
At least, that's what oldcoders comments said.
Let me fix it.

File: code/game/objects/items/weapons/stunbaton.dm [hunk 5]
```diff
@@ -159,38 +159,32 @@
...
 		L.Stun(stunforce)
 		L.apply_effect(10, STUTTER, 0)
 		L.Knockdown(stunforce)
 
-		L.visible_message("<span class='danger'>[L] has been stunned with [src] by [user]!</span>",\
-			"<span class='userdanger'>You have been stunned with [src] by [user]!</span>",\
-			self_drugged_message="<span class='userdanger'>[user]'s [src.name] sucks the life right out of you!</span>")
+		L.visible_message("<span class='danger'>\The [L] has been stunned with \the [src] by [user]!</span>",\
+			"<span class='userdanger'>You have been stunned with \the [src] by \the [user]!</span>",\
+			self_drugged_message="<span class='userdanger'>\the [user]'s [src] sucks the life right out of you!</span>")
```
Comment (DamianX) [inline code/game/objects/items/weapons/stunbaton.dm]: `\The`

**Comment by DamianX** [comment]
So wait, missing on help intent was a bug?

**Comment by ShiftyRail** [comment]
@DamianX I believe so. At any rate, it didn't seem to be intended : missing checked for `hit = -1` and `hit = -1` was only defined if you were either on harm or if the baton was off (for what purpose, I honestly don't know.)

Be advised that the code here is broken, I'm fixing it.

File: code/modules/mob/mob_helpers.dm [hunk 6]
```diff
@@ -14,6 +14,9 @@
 	if(isStunned() || restrained())
 		return 1
 
+/mob/proc/isHandcuffed() // Simple mobs don't have hands ergo aren't cuffed
```
Comment (ShiftyRail) [inline code/modules/mob/mob_helpers.dm]: Sorry if this is out of scope. It was the best way I found to do the checks.

Comment (PJB3005) [reply code/modules/mob/mob_helpers.dm]: You probably want to use `restrained` instead of defining a new proc.

**Comment by ShiftyRail** [comment]
So after all this : 
- Sec can now stun people and harm them. Also works for borgs.
- No chance for missing on help (or disarm/grab) intent, as this was a bug.
- A sound properly plays when attacking people.

I wanted to have a sound for when you miss, as well as to make stunned/restrained/anything people unmissable, but that would require a small scale refactor of `human/attacked_by` so that is for another day.
